src/inmanta/deploy/state.py

```diff
@@ -210,66 +215,193 @@ def are_dependencies_compliant(self, resource: ResourceIdStr) -> bool:
+                else:
+                    # Cache is out of date. Clear cache item.
+                    del known_blockers_cache[resource]
+
+            if blocked_dependency is None:
+                # Perform more expensive call by traversing all requirements of resource.
+                blocked_dependency = next(
+                    (r for r in self.requires.get(resource, set()) if self.resource_state[r].blocked is BlockedStatus.YES), None
+                )
+
+            if blocked_dependency is None:
+                return False
+
+            if blocked_dependency:
+                # Resource is blocked, because a dependency is blocked.
+                known_blockers_cache[resource] = blocked_dependency
+
+            if blocked_dependency in is_blocked:
+                # We know the root is undefined, hard block
+                is_blocked.add(resource)
+
+            # Unblock resource
+            self._block_resource_transitive(resource)
+            out_blocked.add(resource)
+            return True
+
+        for resource_id in verify_blocked:
+            # verify all potential blockers
+            blocked = update_blocked_status_to_blocked(resource_id)
+            # Keep track of this that have been blocked
+            if blocked:
+                propagate_blocked_work.append(resource_id)
+
+        # Propagate blockers once more
+        # We iterate differently, because the parents blocked status is relevant
+        propagate_blocked_work = list(verify_blocked)
+        while propagate_blocked_work:
+            already_blocked: ResourceIdStr = propagate_blocked_work.pop()
+            is_hard_block = already_blocked in is_blocked
+            for to_be_blocked in provides_view.get(already_blocked, set()):
+                if is_hard_block:
+                    is_blocked.add(to_be_blocked)
+                my_state = self.resource_state[to_be_blocked]
+                if my_state.blocked is not BlockedStatus.YES:
+                    continue
+                else:
+                    self._block_resource_transitive(to_be_blocked)
+                    out_blocked.add(to_be_blocked)
+                    propagate_blocked_work.append(to_be_blocked)
 
+        # 2. forward propagate the unblocked
         def update_blocked_status(resource: ResourceIdStr) -> bool:
             """
             Check if the deployment of the given resource is still blocked and mark it as unblocked if it is.
+            check one step over the requires relation, assume all states to hold.
+            if they don't we will pass this node again
 
             :return: True iff the given resource was unblocked.
             """
-            if self.resource_state[resource].blocked is not BlockedStatus.YES:
+            if resource in is_blocked:
+                # Is blocked for a know root
+                return False
+
+            my_state = self.resource_state[resource]
+            if my_state.blocked is not BlockedStatus.YES:
```

Comment:
Do we also want to unblock `BlockedStatus.TRANSIENT` ?

Reply:
@sanderr I don't think so?

Reply:
1. We must set `TRANSIENT -> YES` when applicable
2. Ideally, `TRANSIENT` remains `TRANSIENT` when 1 does not apply. The current implementation might occasionally set `TRANSIENT -> YES -> NO` e.g. if a resource got a new dependency that happens to be in `verify_unblocked`. This is however acceptable imo.